File: DSAlgorithms/patterns/2_pointers/9_quad_sum.cpp
#include <bits/stdc++.h>
class quad_sum_target {
    public:
        static std::vector<std::vector<int>> search_quad(std::vector<int> &arr, int target)
        {
            std::sort(arr.begin(), arr.end());
            std::vector<std::vector<int>> quadruplets;
            for(int i = 0; i < arr.size() -3; i++)
            {
                if(i > 0 && arr[i] == arr[i - 1])
                    continue;
                for(int j = i + 1; j < arr.size() -2; j++)
                {
                    if(j > i + 1 && arr[j] == arr[j - 1])
                        continue;
                    search_pairs(arr, target, i, j, quadruplets);
                }
            }
            return quadruplets;
        }
    private: 
        static void search_pairs(const std::vector<int> &arr, int target_sum, int first, int second, std::vector<std::vector<int>> & quadruplets)
        {
            int left = second + 1;
            int right = arr.size() -1;
            while(left < right)
            {
                int sum = arr[first] + arr[second] + arr[left] + arr[right];
                if(sum == target_sum)
                {
                    quadruplets.push_back({arr[first], arr[second], arr[left], arr[right]});
                    left++;
                    right--;
                    while(left < right && arr[left] == arr[left - 1])
                        left++;
                    while(left < right && arr[right] == arr[right + 1])
                        right--;
                }
                else if(sum < target_sum)
                    left++;
                else
                    right--;
            }
        }
};
int main(int argc, char *argv[])
{
    std::vector<int> vec = {4, 1, 2, -1, 1, -3};
    auto result = quad_sum_target::search_quad(vec, 1);
    for(auto vec : result)
    {
        std::cout<<"[";
        for(auto num : vec)
            std::cout<<num<<" ";
        std::cout<<"]";
    }
    std::cout<<'\n';

    vec = {2, 0, -1, 1, -2, 2};
    result = quad_sum_target::search_quad(vec, 2);
    for(auto vec : result)
    {
        std::cout<<"[";
        for(auto num : vec)
            std::cout<<num<<" ";
        std::cout<<"]";
    }
    std::cout<<'\n';
}

File: DSAlgorithms/patterns/2_pointers/9_quad_sum.py
def search_quad(arr, target):
    arr.sort()
    quadruplets = []
    for i in range(0, len(arr) -3):
        if i > 0 and arr[i] == arr[i - 1]:
            continue
        for j in range(i + 1, len(arr) - 2):
            if j > i + 1 and arr[j] == arr[j - 1]:
                continue
            search_pairs(arr, target, i, j, quadruplets)
    return quadruplets

def search_pairs(arr, target_sum, first, second, quadruplets):
    left = second + 1
    right = len(arr) - 1
    while(left < right):
        sum = arr[first] + arr[second] + arr[left] + arr[right]
        if sum == target_sum:
            quadruplets.append([arr[first], arr[second], arr[left], arr[right]])
            left += 1
            right -= 1
            while(left < right and arr[left] == arr[left -1]):
                left += 1
            while(left < right and arr[right] == arr[right + 1]):
                right -= 1
        elif sum < target_sum:
            left += 1
        else:
            right -= 1
def main():
    print(search_quad([4, 1, 2, -1, 1, -3], 1))
    print(search_quad([2, 0, -1, 1, -2, 2], 2))
main()
